rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,693 @@
+ * @param[in]  filter_taps          A pointer to a lowpass filter used for further decimation.
+ * @param[in]  dm_rate              Decimation rate.
+ * @param[in]  samples_per_antenna  The number of samples per antenna in the original set of
+ *                                  samples.
+ * 
+ * This function performs a parallel version of filtering+downsampling on the GPU to be able
+ * process data in realtime. This algorithm will use 1 GPU thread per filter tap if there are less
+ * than 1024 taps for all filters combined. Only works with power of two length filters, or a
+ * filter that is zero padded to a power of two in length. This algorithm takes one or more 
+ * baseband datasets corresponding to each RX frequency and filters each one using a single lowpass
+ * filter before downsampling.
+ * 
+ *   gridDim.x - The number of decimated output samples for one frequency set.
+ *   gridDim.y - Total number of antennas.
+ *   gridDim.z - Total number of frequency data sets.
+ *
+ *   blockIdx.x - Decimated output sample index.
+ *   blockIdx.y - Antenna index.
+ *   blockIdx.z - Frequency dataset index.
+ *
+ *   blockDim.x - Number of filter taps in the lowpass filter.
+
+ *   threadIdx.x - Filter tap indices.
+ */
+__global__ void lowpass_decimate1024(cuComplex* original_samples,
+  cuComplex* decimated_samples,
+  cuComplex* filter_taps, uint32_t dm_rate,
+  uint32_t samples_per_antenna) {
+
+  // Since number of filter taps is calculated at runtime and we do not want to hardcode
+  // values, the shared memory can be dynamically initialized at invocation of the kernel.
+  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared
+
+  extern __shared__ cuComplex filter_products[];
+
+  auto frequency_dataset_offset = blockIdx.z * (gridDim.x * dm_rate);
+  
+  auto antenna_num = blockIdx.y;
+  auto antenna_offset = antenna_num * samples_per_antenna;
+
+  auto dec_sample_num = blockIdx.x;
+  auto dec_sample_offset = dec_sample_num * dm_rate;
+
+  auto tap_offset = threadIdx.x;
+
+  // If an offset should extend past the length of samples per antenna
+  // then zeroes are used as to not segfault or run into the next buffer.
+  // output samples using these will be discarded as to not introduce edge effects
+  cuComplex sample;
+  if ((dec_sample_offset + threadIdx.x) >= samples_per_antenna) {
+    sample = make_cuComplex(0.0f,0.0f);
+  }
+  else {
+    auto final_offset = frequency_dataset_offset + antenna_offset + dec_sample_offset + threadIdx.x;
+    sample = original_samples[final_offset];
+  }
+
+
+  filter_products[tap_offset] = cuCmulf(sample,filter_taps[tap_offset]);
+  // Synchronizes all threads in a block, meaning 1 output sample per rx freq
+  // is ready to be calculated with the parallel reduce
+  __syncthreads();
+
+  auto calculated_output_sample = parallel_reduce(filter_products, tap_offset);
+
+  // When decimating, we go from one set of samples for each antenna
```

Comment:
We think this comment needs updating now that this is a lowpass and the input samples ant0, freq0; ant1; freq0...antx, freq0; ant0; freq1 ...